Spn smearing

theHoHsiao opened the pull request:
We implemented the Wuppertal smearing and APE smearing for both connected and disconnected measurements.

theHoHsiao committed:
add code for APE and Wuppertal smearing (does not compile - old complex numbers)

fzierler committed:
rename source-and-sink smearing measurement code

fzierler committed:
perform all sink smearing levels at once

fzierler committed:
add smeared volume sources, smearing for disconnected pieces

fzierler committed:
add example input file for meson measurements with smearing

fzierler committed:
switch over to new complex numbers

fzierler committed:
add missing call to smeared disconnected measurement

fzierler commented:
Hi, as discussed with Paul, I found that the code here is actually slower than the one developed on top of the old HiRep code (using the old complex numbers). Specifically, I tested this code here and compared it to the one at https://github.com/fzierler/hirep-smear on DiaL3 for the connected measurements on a 64x16^3 lattice using the same input file and same flags in the Makefile. I find that the code here is about 50% slower. I don't know what could have caused this behaviour.

fzierler commented:
Hi, I also tried to test this code for a source position different from (0,0,0,0). However, this leads to a failure of the Inverter on some of the configurations. Typically, this depends on the MPI partitioning and the specific machine (or maybe compiler). Further testing shows that this appears even without smearing in the conventional code for point sources. 

I suggest we remove the option to place the source at an arbitrary point for the smearing code.

fzierler commented:
Hi, currently this code does not measure the disconnected pieces because a function call to this measurement was lost, when I ported the code to the recent version of HiRep. I have fixed this by now and created PR for Paul's branch here: https://github.com/theHoHsiao/HiRep_smear/pull/1

edbennett commented:
Hi both, just returning to this now—should this PR be merged in its current state, or should it wait until theHoHsiao#1 has been merged?

fzierler commented:
Hi Ed, please wait until https://github.com/theHoHsiao/HiRep_smear/pull/1 is merged.

theHoHsiao committed:
Merge pull request #1 from fzierler/spn_smearing_missing_disconnected

add missing call to smeared disconnected measurement

edbennett reviewed (changes requested):
Many thanks for this.

Is there any chance please that for each function that is introduced, a comment could be added at the top giving
1. A one-sentence description of what the function does, in more words than the function name
2. If not obvious, the meaning/purpose of any parameters.
3. A reference to a paper or textbook, including equation numbers, for definintions of this.

E.g. `APE_smearing` might get the comment:
```
/* Updates the global APE smeared gauge field
   by performing Nsmear steps of APE smearing.
   Each step is Eq. (5) of 0801.1165;
   smear_val here is $\alpha$ in the convention of the paper. */
```

This will hopefully mean that future people working on the code have a much easier time working out what each function does. I realise that most of the existing code doesn't do this, but I think it's a difficulty we've all had when reading it, so hopefully we can move in the direction of sparing others the same frustration.

theHoHsiao commented:
Hi Ed,

Sure! The code was written long ago when I was still a master's degree student. I will write some descriptions for the functions I implemented.

Fabian, maybe after the Christmas break, could you take care of the parts related to the disconnected calculations? I should finish writing the descriptions by that time.

edbennett commented:
Thanks! Tagging @fzierler in case he's not subscribed to the issue.

theHoHsiao committed:
add comments for Wuppertal smearing and helper function

1. write descriptions for the function
2. define helper functions to increase readability
3. remove unused arguments `int color` in `smearing_function`

theHoHsiao committed:
Update .gitignore

theHoHsiao committed:
refine comments

theHoHsiao committed:
add comments

theHoHsiao committed:
refine comments

theHoHsiao committed:
refine comments

theHoHsiao commented:
Hi @edbennett and @fzierler,

Happy New Year! I hope you had a great holiday.

I have added some comments for the relevant functions.
I also refactor the function, `measure_smearing_source_sink`, with two helper functions, `create_smear_source_and_measure` and `smear_sink_and_measure`, to improve the readability.

Please have a look at it in your own time. @fzierler, I am leaving you the comments for the disconnected part. Thanks.
Feel free to let me know if they still need any improvements.

edbennett commented on LibHR/Utils/suN_utils.c:
```diff
@@ -447,3 +447,146 @@ void covariant_project_to_suNg(suNg *u)
 }
+
+
+void cooling_SPN(suNg* g_out, suNg* g_in, suNg* g_tilde, int cooling){
```
What does this function do? Are $g$, $\tilde{g}$, $B$, $S$, … defined in a paper that we can cite here?

theHoHsiao replied:
I could not find any reference for the cooling method for the Sp(2N) gauge group. Do you recall any papers you took as a reference when computing the glueballs?

edbennett replied:
I wasn't heavily involved in that computation; @dvadacchino do you remember this, please?

dvadacchino replied:
Cooling for Sp(2N) follows the same logic as for SU(N). The logic is explained at pages 9-10 of https://arxiv.org/pdf/hep-lat/0404008, i.e. by using the beta=infinity limit of the cabibbo-marinari algorithm, and then using gram-schmidt for resymplecticization.

theHoHsiao replied:
Okay. I will implement some more comments based on that paper.

theHoHsiao replied:
I have added some comments for `cooling_SPN` and its relevant helper functions: `subgrb`, `subgrb_tau`, `vmxsu2`, and `vmxsu2_tau`.

edbennett commented on LibHR/Utils/suN_utils.c:
```diff
@@ -447,3 +447,146 @@ void covariant_project_to_suNg(suNg *u)
+    }
+}
+
+void subgrb(int i1col, int i2col, suNgfull* B11, suNgfull* C11){
```
What does this function do? Is there a paper we can cite?

dvadacchino replied:
this picks a su(2) submatrix of the kind (A4) in https://arxiv.org/pdf/2010.15781  of an sp(2n) matrix

edbennett commented on LibHR/Utils/suN_utils.c:
```diff
@@ -447,3 +447,146 @@ void covariant_project_to_suNg(suNg *u)
+    vmxsu2(i1col,i2col,B11,A11);
+}
+
+void vmxsu2(int i1, int i2, suNgfull* A, double complex B[4]){
```
What does this function do? Is there a paper we can cite?

dvadacchino replied:
This is vector matrix multiplication for su2 matrices

edbennett commented on LibHR/Utils/suN_utils.c:
```diff
@@ -447,3 +447,146 @@ void covariant_project_to_suNg(suNg *u)
+    }
+}
+
+void subgrb_tau(int n1, int n2, suNgfull* B11, suNgfull* C11){
```
What does this function do? How is it different from `subgrb`? Is there a paper we can cite?

dvadacchino replied:
this picks a su(2) submatrix of the kind (A5) in https://arxiv.org/pdf/2010.15781  of an sp(2n) matrix

edbennett commented on LibHR/Utils/suN_utils.c:
```diff
@@ -447,3 +447,146 @@ void covariant_project_to_suNg(suNg *u)
+    vmxsu2_tau(n1, n2, B11, A11);
+}
+
+void vmxsu2_tau(int n1, int n2, suNgfull* A, complex B[4]){
```
What does this function do? How is it different from `vmxsu2`? Is there a paper we can cite?

edbennett commented on LibHR/Observables/polyakov.c:
```diff
@@ -458,4 +458,252 @@ void polyakov()
+#undef _MIN
+
+
+void polyakov_APE() {
```
Move this function above the two `#undef` lines—these `#define`s are intended to apply for the entire file; having the `#undef`s hidden in the middle makes it unclear that this header doesn't introduce side-effects.

edbennett replied:
What does this function do? Is there a paper we can cite?

edbennett commented on LibHR/Observables/meson_measurements.c:
```diff
@@ -761,3 +761,231 @@ void measure_formfactor_fixed(int ti, int tf, int dt, int nm, double *m, int n_m
+// So far, only implement the measurement for mesons at rest
+// The sources are not localised on any lattice site but spread out
+/* Smearing for disconnected pieces for singlets (FZ Jan. 2024)*/
+void measure_spectrum_discon_semwall_smeared_single_inversion(int nm, double* m, int nhits, int conf_num, double precision, double epsilon_source, int Nsmear_source, double APE_epsilon, int APE_N, int N_diff){
```
Can this comment go into more detail (and possibly be formatted like the others introduced in this PR)?

theHoHsiao replied:
@fzierler Fabian, can I leave this to you?

edbennett commented on LibHR/Observables/meson_measurements.c:
```diff
@@ -761,3 +761,231 @@ void measure_formfactor_fixed(int ti, int tf, int dt, int nm, double *m, int n_m
+}
+
+void measure_smearing_ss(int t, int x, int y, int z, int nm, double* m, int n_mom, int nhits, int conf_num, double precision, double epsilon_source, int Nsmear_source, double epsilon_sink, int Nsmear_sink, double APE_epsilon, int APE_N, int N_diff){
+  /**
```
Can this comment include a reference defining the procedure implemented and variable names used (and possibly be formatted like the others introduced in this PR)?

edbennett commented on LibHR/Observables/meson_measurements.c:
```diff
@@ -761,3 +761,231 @@ void measure_formfactor_fixed(int ti, int tf, int dt, int nm, double *m, int n_m
+}
+
+
+void measure_smearing_source_sink(int t, int x, int y, int z, int nm, double* m, int n_mom, int nhits, int conf_num, double precision, double epsilon_source, int Nsmear_source_max, double epsilon_sink, int Nsmear_sink, double APE_epsilon, int APE_N, int N_diff){
```
Can this comment include a reference defining the procedure implemented and variable names used (and possibly be formatted like the others introduced in this PR)?

edbennett commented on LibHR/Observables/meson_measurements.c:
```diff
@@ -761,3 +761,231 @@ void measure_formfactor_fixed(int ti, int tf, int dt, int nm, double *m, int n_m
+}
+
+
+void smear_sink_and_measure(spinor_field* prop, spinor_field* source, int nm, double* m, int t, int n_mom, int Nsmear_source, double epsilon_sink, int Nsmear_sink, int N_diff, int APE_N, int conf_num, char* label) {
```
Can this comment include a reference defining the procedure implemented and variable names used (and possibly be formatted like the others introduced in this PR)?

edbennett commented on LibHR/Observables/meson_measurements.c:
```diff
@@ -761,3 +761,231 @@ void measure_formfactor_fixed(int ti, int tf, int dt, int nm, double *m, int n_m
+/* Updates for Wuppertal smearing by HH in 2020*/
+
+void create_smear_source_and_measure(spinor_field* source, spinor_field* prop, int t, int x, int y, int z, int nm, double* m, int n_mom, double epsilon_source, int Nsmear_source, int APE_N,int conf_num, char* label) {
+  /**
```
Can this comment include a reference defining the procedure implemented and variable names used (and possibly be formatted like the others introduced in this PR)?

edbennett commented on LibHR/Observables/measure_mesons.c:
```diff
@@ -623,3 +623,406 @@ void print_mesons(meson_observable *mo, double norm, int conf, int nm, double *m
+    free_spinor_field_f(smeared_psi);
+}
+
+void smeared_propagator_volume_with_APE(spinor_field* psi, int nm , double epsilon){
```
What does this function do? Is there a paper we can cite?

theHoHsiao replied:
@fzierler Fabian, can I leave this to you?

edbennett commented on LibHR/Observables/measure_mesons.c:
```diff
@@ -623,3 +623,406 @@ void print_mesons(meson_observable *mo, double norm, int conf, int nm, double *m
+
+
+// FZ (2024): Smeared propagator with a single source
+void smeared_propagator_volume(spinor_field* psi, int nm , double epsilon){
```
What does this function do? Is there a paper we can cite?

theHoHsiao replied:
@fzierler Fabian, can I leave this to you?

edbennett commented on LibHR/Observables/avr_plaquette.c:
```diff
@@ -416,3 +416,182 @@ double complex avr_plaquette_wrk()
+
+}
+
+void full_plaquette_APE()
```
What does this function do? Is there a paper we can cite?

theHoHsiao replied:
I added some comments. It computes the normal plaquette for the APE-smeared gauge links. Do we need a paper for this?

edbennett replied:
Ideally, I'd suggest something like "The full plaquette $P$ as defined in Eq. (123) of arXiv:1234.5678, for the APE smeared gauge field u_gauge_APE".

edbennett commented on LibHR/Observables/avr_plaquette.c:
```diff
@@ -416,3 +416,182 @@ double complex avr_plaquette_wrk()
+}
+
+
+double avr_spacial_plaquette_APE()
```
What does this function do? Is there a paper we can cite?

edbennett replied:
Is there a reason that [`spacial`](https://en.wiktionary.org/wiki/spacial) here is used instead of [`spatial`](https://en.wiktionary.org/wiki/spatial)?

edbennett commented on LibHR/Observables/avr_plaquette.c:
```diff
@@ -416,3 +416,182 @@ double complex avr_plaquette_wrk()
+#endif
+}
+
+void cplaq_APE(complex *ret,int ix,int mu,int nu)
```
What does this function do? Is there a paper we can cite?

edbennett commented on LibHR/Observables/avr_plaquette.c:
```diff
@@ -416,3 +416,182 @@ double complex avr_plaquette_wrk()
 }
+
+
+double plaq_APE(int ix,int mu,int nu)
```
What does this function do? Is there a paper we can cite?

edbennett commented on LibHR/Observables/meson_measurements.c:
```diff
@@ -761,3 +761,231 @@ void measure_formfactor_fixed(int ti, int tf, int dt, int nm, double *m, int n_m
+}
+// So far, only implement the measurement for mesons at rest
+// The sources are not localised on any lattice site but spread out
+/* Smearing for disconnected pieces for singlets (FZ Jan. 2024)*/
```
Can this comment go into more detail (and possibly be formatted like the others introduced in this PR)?

theHoHsiao replied:
@fzierler Fabian, can I leave this to you?

edbennett reviewed (changes requested):
Many thanks Paul, those comments are really helpful. I've tagged the remaining functions that I think would benefit from some explanation.

theHoHsiao commented:
Hi Ed,

Thanks for the review. As you pointed out, there are some spelling typos, and I forgot to say that I am still working on the APE smearing part. I will improve them soon.

theHoHsiao committed:
correct typo

theHoHsiao committed:
add comments

theHoHsiao committed:
improving comments

theHoHsiao committed:
add comments for `cooling_SPN` and helper functions
